---
title: "bcell processing"
author: "Jin Gyu Cheong"
date: "2022-08-22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

BALF N (Naive)
BALF R (recovered)

## data loading
```{r load data}
library(Signac)
library(Seurat)
library(EnsDb.Mmusculus.v79)
library(BSgenome.Mmusculus.UCSC.mm10)
library(stringr)
nset.seed(880619)

countsN <- Read10X_h5("./MA10_BALF_N/filtered_feature_bc_matrix.h5")
fragpathN <- "./MA10_BALF_N/atac_fragments.tsv.gz"

countsR <- Read10X_h5("./MA10_BALF_R/filtered_feature_bc_matrix.h5")
fragpathR <- "./MA10_BALF_R/atac_fragments.tsv.gz"


# get gene annotations for hg38
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
ucsc.levels <- str_replace(string=paste("chr",seqlevels(annotation),sep=""), pattern="chrMT", replacement="chrM")
seqlevels(annotation) <- ucsc.levels
# seqlevelsStyle(annotation) <- "UCSC"  doesnt work

# create a Seurat object containing the RNA adata
repN <- CreateSeuratObject(
  counts = countsN$`Gene Expression`,
  assay = "RNA"
)

repR <- CreateSeuratObject(
  counts = countsR$`Gene Expression`,
  assay = "RNA"
)

# create ATAC assay and add it to the object

repN[["ATAC"]] <- CreateChromatinAssay(
  counts = countsN$Peaks,
  sep = c(":", "-"),
  fragments = fragpathN,
  annotation = annotation
)

repR[["ATAC"]] <- CreateChromatinAssay(
  counts = countsR$Peaks,
  sep = c(":", "-"),
  fragments = fragpathR,
  annotation = annotation
)
```

## QC for Naive
```{r pressure, echo=FALSE}
DefaultAssay(repN) <- "ATAC"

repN <- NucleosomeSignal(repN)
repN <- TSSEnrichment(repN)

VlnPlot(
  object = repN,
  features = c("nCount_RNA", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal"),
  ncol = 4,
  pt.size = 0
)

# filter out low quality cells
repN <- subset(
  x = repN,
  subset = nCount_ATAC < 150000 &
    nCount_RNA < 20000 &
    nCount_ATAC > 1000 &
    nCount_RNA > 1000 &
    nucleosome_signal < 2 &
    TSS.enrichment > 1
)
repN
```

## QC for Recovered
```{r pressure, echo=FALSE}
DefaultAssay(repR) <- "ATAC"

repR <- NucleosomeSignal(repR)
repR <- TSSEnrichment(repR)

VlnPlot(
  object = repR,
  features = c("nCount_RNA", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal"),
  ncol = 4,
  pt.size = 0
)

# filter out low quality cells
repR <- subset(
  x = repR,
  subset = nCount_ATAC < 100000 &
    nCount_RNA < 25000 &  # 
     nCount_ATAC > 1000 &
    nCount_RNA > 1000 &
    nucleosome_signal < 2 &
    TSS.enrichment > 1
)

```

## peak calling
```{r peak calling}
# call peaks using MACS2
peaksN <- CallPeaks(repN, macs2.path = "/Users/jlab/miniconda3/bin/macs2")

# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaksN <- keepStandardChromosomes(peaksN, pruning.mode = "coarse")
peaksN <- subsetByOverlaps(x = peaksN, ranges = blacklist_mm10, invert = TRUE)

# quantify counts in each peak
#macs2_counts2 <- FeatureMatrix(
#  fragments = Fragments(rep2),
#  features = peaks2,
#  cells = colnames(rep2)
#)

# create a new assay using the MACS2 peak set and add it to the Seurat object
#rep2[["peaks"]] <- CreateChromatinAssay(
#  counts = macs2_counts2,
#  fragments = fragpath2,
#  annotation = annotation
#)

# call peaks using MACS2
peaksR <- CallPeaks(repR, macs2.path = "/Users/jlab/miniconda3/bin/macs2")

# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaksR <- keepStandardChromosomes(peaksR, pruning.mode = "coarse")
peaksR <- subsetByOverlaps(x = peaksR, ranges = blacklist_mm10, invert = TRUE)

# quantify counts in each peak
#macs2_counts3 <- FeatureMatrix(
#  fragments = Fragments(rep3),
#  features = peaks3,
#  cells = colnames(rep3)
#)

# create a new assay using the MACS2 peak set and add it to the Seurat object
#rep3[["peaks"]] <- CreateChromatinAssay(
#  counts = macs2_counts3,
#  fragments = fragpath3,
#  annotation = annotation
#)
```


# merging two ATAC
```{r echo=F, tidy=T}
library(tidyverse)
library(GenomicRanges)
library(future)


gr.repN <- makeGRangesFromDataFrame(peaksN)
gr.repR <- makeGRangesFromDataFrame(peaksR)

combined.peaks <- GenomicRanges::reduce(x = c(peaksN, peaksR))

peakwidths <- width(combined.peaks)
combined.peaks <- combined.peaks[peakwidths  < 10000 & peakwidths > 20]
combined.peaks
saveRDS(combined.peaks, "./BALF_combined.peak.rds")

md.repN <- read.table(
  file = "./MA10_BALF_N/per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)[-1, ] # remove the first row

md.repR <- read.table(
  file = "./MA10_BALF_R/per_barcode_metrics.csv",
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
)[-1, ] # remove the first row


md.repN <- md.repN[Cells(repN), ]
md.repR <- md.repR[Cells(repR), ]

frags.repN <- CreateFragmentObject(
  path = "./MA10_BALF_N/atac_fragments.tsv.gz",
  cells = rownames(md.repN))

frags.repR <- CreateFragmentObject(
  path = "./MA10_BALF_R/atac_fragments.tsv.gz",
  cells = rownames(md.repR))


repN.counts <- FeatureMatrix(
  fragments = frags.repN,
  features = combined.peaks,
  cells = rownames(md.repN)
)

repR.counts <- FeatureMatrix(
  fragments = frags.repR,
  features = combined.peaks,
  cells = rownames(md.repR)
)


repN_assay <- CreateChromatinAssay(repN.counts, fragments = frags.repN)
repN_2 <- CreateSeuratObject(repN_assay, assay = "ATAC", meta.data=md.repN)

repR_assay <- CreateChromatinAssay(repR.counts, fragments = frags.repR)
repR_2 <- CreateSeuratObject(repR_assay, assay = "ATAC", meta.data=md.repR)


repN_2$dataset <- 'Naive'
repR_2$dataset <- 'Recovered'

combined <- merge(
  x = repN_2,
  y = repR_2,
  add.cell.ids = c("N", "R")
)


DimPlot(combined, group.by = 'dataset', pt.size = 0.1)
```


# merging RNAseq object
```{r}
DefaultAssay(repN) <- "RNA"
DefaultAssay(repR) <- "RNA"
merged_rna <- merge(x=repN, y = repR, add.cell.ids = c("N", "R"))
```



# add RNA assay to the ATAC
get cell indices shared in both data type 
intersect cell indices from ATAC and RNA
add merged RNA assay to ATACseq assay
```{r}

intersect(Cells(merged_rna),Cells(combined)) -> idx
combined <- subset(combined, cells = idx)
merged_rna <- subset(merged_rna, cells = idx)

combined[["RNA"]] <- merged_rna[["RNA"]]


# get gene annotations for hg38
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
seqlevelsStyle(annotation) <- "UCSC" # doesnt work

macs2_counts <- FeatureMatrix(
  fragments = Fragments(combined),
  features = combined.peaks,
  cells = colnames(combined)
)

combined[["peaks"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = Fragments(combined),
  annotation = annotation
)

saveRDS(combined, "./MA10_BALF_combined.RDS")
```


# normalization and UMAP
```{r RNAseq data}
#RNA UMAP
DefaultAssay(combined) <- "RNA"
combined<- NormalizeData(object = combined)
combined<- FindVariableFeatures(object = combined)
combined<- ScaleData(object = combined)
combined<- RunPCA(combined, verbose = FALSE)
combined<- RunUMAP(combined, dims = 1:30, verbose = FALSE, reduction.name = "RNA_umap" )

# SCT normalization and generate UMAP
DefaultAssay(combined) <- "RNA"
combined <- SCTransform(combined)
combined <- RunPCA(combined)
combined <- RunUMAP(combined, dims = 1:30, verbose = FALSE,reduction.name = "SCT_umap")

# ATACseq pre processing and generate UMAP
DefaultAssay(combined) <- "peaks"
combined <- RunTFIDF(combined)
combined <- FindTopFeatures(combined, min.cutoff = 20)
combined <- RunSVD(combined)
combined <- RunUMAP(combined, dims = 2:50, reduction = 'lsi', reduction.name = "ATAC_umap")

```

WNN integration
```{r}
# build a joint neighbor graph using both assays
combined <- FindMultiModalNeighbors(
  object = combined,
  reduction.list = list("pca", "lsi"), 
  dims.list = list(1:50, 2:40),
  modality.weight.name = "RNA.weight",
  verbose = TRUE
)

# build a joint UMAP visualization
combined <- RunUMAP(
  object = combined,
  nn.name = "weighted.nn",
  assay = "RNA",
  verbose = TRUE,  reduction.name = "umap_WNN"
)
```


# clustering
```{r}
DefaultAssay(combined) <- "RNA" 
combined<- FindNeighbors(combined, dims = 1:30, verbose = FALSE)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "RNA_snn")
combined<- FindClusters(combined, verbose = FALSE, graph.name = "RNA_snn", resolution = 0.6)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "RNA_snn", resolution = 0.4)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "RNA_snn", resolution = 0.2)
#SCT UMAP
DefaultAssay(combined) <- "SCT"
combined<- FindNeighbors(combined, dims = 1:30, verbose = FALSE)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "SCT_snn", resolution = 0.8)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "SCT_snn", resolution = 0.6)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "SCT_snn", resolution = 0.4)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "SCT_snn", resolution = 0.2)


#LSI UMAP
DefaultAssay(combined) <- "peaks"
combined<- FindNeighbors(combined, dims = 2:30, verbose = FALSE, reduction = "lsi")
combined<- FindClusters(combined, verbose = FALSE, graph.name = "peaks_snn")
combined<- FindClusters(combined, verbose = FALSE, graph.name = "peaks_snn", resolution = 0.6)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "peaks_snn", resolution = 0.4)
combined<- FindClusters(combined, verbose = FALSE, graph.name = "peaks_snn", resolution = 0.2)

pdf("./BALF_umap_various_clustering.pdf")
DimPlot(combined, reduction = "RNA_umap", group.by = "RNA_snn_res.0.8") + ggtitle("RNA 0.8")
DimPlot(combined, reduction = "RNA_umap", group.by = "RNA_snn_res.0.6") + ggtitle("RNA 0.6")
DimPlot(combined, reduction = "RNA_umap", group.by = "RNA_snn_res.0.4") + ggtitle("RNA 0.4")
DimPlot(combined, reduction = "RNA_umap", group.by = "RNA_snn_res.0.2") + ggtitle("RNA 0.2")

DimPlot(combined, reduction = "umap_SCT", group.by = "SCT_snn_res.0.8") + ggtitle("SCT 0.8")
DimPlot(combined, reduction = "umap_SCT", group.by = "SCT_snn_res.0.6") + ggtitle("SCT 0.6")
DimPlot(combined, reduction = "umap_SCT", group.by = "SCT_snn_res.0.4") + ggtitle("SCT 0.4")
DimPlot(combined, reduction = "umap_SCT", group.by = "SCT_snn_res.0.2") + ggtitle("SCT 0.2")

DimPlot(combined, reduction = "umap_ATAC", group.by = "peaks_snn_res.0.8") + ggtitle("peaks 0.8")
DimPlot(combined, reduction = "umap_ATAC", group.by = "peaks_snn_res.0.6") + ggtitle("peaks 0.6")
DimPlot(combined, reduction = "umap_ATAC", group.by = "peaks_snn_res.0.4") + ggtitle("peaks 0.4")
DimPlot(combined, reduction = "umap_ATAC", group.by = "peaks_snn_res.0.2") + ggtitle("peaks 0.2")

DimPlot(combined, reduction = "umap_WNN", group.by = "RNA_snn_res.0.8") + ggtitle("RNA 0.8 on WNN")
DimPlot(combined, reduction = "umap_WNN", group.by = "SCT_snn_res.0.8") + ggtitle("SCT 0.8 on WNN")
DimPlot(combined, reduction = "umap_WNN", group.by = "peaks_snn_res.0.8") + ggtitle("peaks 0.8 on WNN")
dev.off()

```

# cluster marker genes 

RNA 0.2
```{r}
balfcombined <- combined

DefaultAssay(balfcombined) <- "RNA"
Idents(balfcombined) <- "RNA_snn_res.0.2"

markers <- FindAllMarkers(balfcombined, assay = "SCT", only.pos = T)

openxlsx::write.xlsx(markers, "./BALF_cluster_marker_genes_RNAres02.xlsx")

top10 <- markers %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) 

pdf("./BALF_markers_heatmap_RNAres02.pdf", height = 30, width = 30)
DoHeatmap(balfcombined, features = top10$gene) + NoLegend()
dev.off()


markers_RNAres02 <- read.xlsx("BALF_cluster_marker_genes_RNAres02.xlsx")
top10_RNAres02 <- markers_RNAres02 %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) 
genes <- top10_RNAres02$gene %>% unique()


DefaultAssay(balfcombined) <- "SCT"

plot <- list()
for (i in 1:length(genes)){
  plot[[i]] <- FeaturePlot(balfcombined, genes[i])
}

pdf("./balf_RNAres02_markergene_featureplot.pdf")
plot
dev.off()

```


ATAC 0.2
```{r}
balfcombined <- combined

DefaultAssay(balfcombined) <- "SCT"
Idents(balfcombined) <- "peaks_snn_res.0.2"

markers <- FindAllMarkers(balfcombined, assay = "SCT", only.pos = T)

openxlsx::write.xlsx(markers, "./BALF_cluster_marker_genes_Peaksres02.xlsx")

top10 <- markers %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) 

pdf("./BALF_markers_heatmap_Peaksres02.pdf", height = 30, width = 30)
DoHeatmap(balfcombined, features = top10$gene) + NoLegend()
dev.off()

markers_ATACres02 <- read.xlsx("BALF_cluster_marker_genes_Peaksres02.xlsx")
top10_ATACres02 <- markers_ATACres02 %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) 
genes <- top10_RNAres02$gene %>% unique()

DefaultAssay(balfcombined) <- "SCT"

plot <- list()
for (i in 1:length(genes)){
  plot[[i]] <- FeaturePlot(balfcombined, genes[i])
}

pdf("./balf_ATACres02_markergene_featureplot.pdf")
plot
dev.off()


```


# transfer annotation

## transfer annotation from single cell RNA-seq data
```{r}
scbalf <- readRDS("/Users/jlab/Library/CloudStorage/OneDrive-med.cornell.edu/7_MA10project/sarscov2iavbalf_scRNASeq.rds")
scbalf
```


```{r annotation}

DefaultAssay(scbalf) <- "SCT"
DefaultAssay(balfcombined) <- "SCT"

scbalf <- FindVariableFeatures(scbalf)
scbalf <- SCTransform(scbalf)

# transfer cell type labels from reference to query
transfer_anchors <- FindTransferAnchors(
  reference = scbalf,
  query = balfcombined,
  normalization.method = "SCT",
  reference.reduction = "pca",
  recompute.residuals = FALSE,
  dims = 1:50
)

predictions <- TransferData(
  anchorset = transfer_anchors,
  reference = scbalf,
  refdata = "minor_group",
  weight.reduction = balfcombined[['pca']],
  dims = 1:50
)

balfcombined <- AddMetaData(
  object = balfcombined,
  metadata = predictions$predicted.id,col.name = "minor_group"
)

predictions <- TransferData(
  anchorset = transfer_anchors,
  reference = scbalf,
  refdata = "major_group",
  weight.reduction = balfcombined[['pca']],
  dims = 1:50
)

balfcombined <- AddMetaData(
  object = balfcombined,
  metadata = predictions$predicted.id,col.name = "major_group"
)

predictions <- TransferData(
  anchorset = transfer_anchors,
  reference = scbalf,
  refdata = "cluster_annotation",
  weight.reduction = balfcombined[['pca']],
  dims = 1:50
)

balfcombined <- AddMetaData(
  object = balfcombined,
  metadata = predictions$predicted.id,col.name = "cluster_annotation"
)

```

# celltype annotation
```{r}
combined$celltype <- NA
combined$celltype[combined$peaks_snn_res.0.2 == 0] <- "macrophage cluster 1"
combined$celltype[combined$peaks_snn_res.0.2 == 1] <- "macrophage cluster 2"
combined$celltype[combined$peaks_snn_res.0.2 == 2] <- "CD8 clustser 1"
combined$celltype[combined$peaks_snn_res.0.2 == 3] <- "CD8 clustser 2"
combined$celltype[combined$peaks_snn_res.0.2 == 4] <- "myeloid cells"
combined$celltype[combined$peaks_snn_res.0.2 == 6] <- "CD4 T cells"
combined$celltype[combined$peaks_snn_res.0.2 == 5] <- "B cells"
combined$celltype[combined$peaks_snn_res.0.2 == 7] <- "dendritic cells"
combined$celltype[combined$peaks_snn_res.0.2 == 8] <- "epithelial cells"
```

# remove dobulet using scrublet
```{r}
library(reticulate)
library(tidyverse)
# You can set the conda environment
conda_create("base")
conda_install("base", packages =c("scipy", "matplotlib", "numpy", "scrublet"))
use_condaenv("base", required = T)
```

```{python}

import scrublet as scr
import scipy.io
import matplotlib.pyplot as plt
import numpy as np
import os
def predict_doublets(matrix):
  counts_matrix = scipy.io.mmread(matrix).T.tocsc()
  scrub = scr.Scrublet(counts_matrix, expected_doublet_rate=0.10)
  doublet_scores, predicted_doublets = scrub.scrub_doublets(min_counts=2, min_cells=3, min_gene_variability_pctl=85, n_prin_comps=30)
  return scrub.doublet_scores_obs_
#scrub.plot_histogram()
#scrub.set_embedding('UMAP', scr.get_umap(scrub.manifold_obs_, 10, min_dist=0.3))
#scrub.plot_embedding('UMAP', order_points=True)

matrix = "./MA10_BALF_N/filtered_feature_bc_matrix/matrix.mtx.gz"
counts_matrix = scipy.io.mmread(matrix).T.tocsc()
scrubN = scr.Scrublet(counts_matrix, expected_doublet_rate=0.10)
doublet_scoresN, predicted_doubletsN = scrubN.scrub_doublets(min_counts=2, min_cells=3, min_gene_variability_pctl=85, n_prin_comps=30)

matrix = "./MA10_BALF_R/filtered_feature_bc_matrix/matrix.mtx.gz"
counts_matrix = scipy.io.mmread(matrix).T.tocsc()
scrubR = scr.Scrublet(counts_matrix, expected_doublet_rate=0.10)
doublet_scoresR, predicted_doubletsR = scrubR.scrub_doublets(min_counts=2, min_cells=3, min_gene_variability_pctl=85, n_prin_comps=30)

```


```{r}
barcodeN <- read.table("./MA10_BALF_N/filtered_feature_bc_matrix/barcodes.tsv.gz")
doubletN <- py$scrubN$doublet_scores_obs_
doubletN <- as.vector(doubletN)
barcodeN$dbl_score <- doubletN
barcodeN$is.dbl <- py$predicted_doubletsN
colnames(barcodeN) <- c("barcodes","dbl_score","is.dbl")

barcodeR <- read.table("./MA10_BALF_R/filtered_feature_bc_matrix/barcodes.tsv.gz")
doubletR <- py$scrubR$doublet_scores_obs_
doubletR <- as.vector(doubletR)
barcodeR$dbl_score <- doubletR
barcodeR$is.dbl <- py$predicted_doubletsR
colnames(barcodeR) <- c("barcodes","dbl_score","is.dbl")
```


# linking peaks to genes
```{r peak to genes}
DefaultAssay(pbmc) <- "peaks"

# first compute the GC content for each peak
pbmc <- RegionStats(pbmc, genome = BSgenome.Hsapiens.UCSC.hg38)

# link peaks to genes
pbmc <- LinkPeaks(
  object = pbmc,
  peak.assay = "peaks",
  expression.assay = "SCT",
  genes.use = c("LYZ", "MS4A1")
)

idents.plot <- c("B naive", "B intermediate", "B memory",
                 "CD14 Mono", "CD16 Mono", "CD8 TEM", "CD8 Naive")

p1 <- CoveragePlot(
  object = pbmc,
  region = "MS4A1",
  features = "MS4A1",
  expression.assay = "SCT",
  idents = idents.plot,
  extend.upstream = 500,
  extend.downstream = 10000
)
```


# Find differential peaks and enrich motif
## Finding overrepresented motifs
```{r}
BiocManager::install("JASPAR2020")
BiocManager::install("TFBSTools")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
BiocManager::install("patchwork")

library(Signac)
library(Seurat)
library(JASPAR2020)
library(TFBSTools)
library(BSgenome.Mmusculus.UCSC.mm10)
library(patchwork)
set.seed(880619)

combined <- readRDS("./merged_BM_multiome_obj.rds")

# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
balfcombined <- AddMotifs(
  object = balfcombined,
  genome = BSgenome.Mmusculus.UCSC.mm10,
  pfm = pfm
)

da_peaks <- FindMarkers(
  object = balfcombined,
  ident.1 = 'Pvalb',
  ident.2 = 'Sst',
  only.pos = TRUE,
  test.use = 'LR',
  min.pct = 0.05,
  latent.vars = 'nCount_peaks'
)

# get top differentially accessible peaks
top.da.peak <- rownames(da_peaks[da_peaks$p_val < 0.005, ])

# test enrichment
enriched.motifs <- FindMotifs(
  object = mouse_brain,
  features = top.da.peak
)

MotifPlot(
  object = mouse_brain,
  motifs = head(rownames(enriched.motifs))
)

mouse_brain <- RunChromVAR(
  object = mouse_brain,
  genome = BSgenome.Mmusculus.UCSC.mm10
)

DefaultAssay(mouse_brain) <- 'chromvar'

# look at the activity of Mef2c
p2 <- FeaturePlot(
  object = mouse_brain,
  features = "MA0497.1",
  min.cutoff = 'q10',
  max.cutoff = 'q90',
  pt.size = 0.1
)
p1 + p2

```


# Chromvar :: Computing motif activities
We can also compute a per-cell motif activity score by running chromVAR. This allows us to visualize motif activities per cell, and also provides an alternative method of identifying differentially-active motifs between cell types.
ChromVAR identifies motifs associated with variability in chromatin accessibility between cells. See the chromVAR paper for a complete description of the method.
```{r}

DefaultAssay(balfcombined) <- "peaks"

balfcombined <- RunChromVAR(
  object = balfcombined,
  genome = BSgenome.Mmusculus.UCSC.mm10
)

DefaultAssay(mouse_brain) <- 'chromvar'

# look at the activity of Mef2c
p2 <- FeaturePlot(
  object = mouse_brain,
  features = "MA0497.1",
  min.cutoff = 'q10',
  max.cutoff = 'q90',
  pt.size = 0.1
)
p1 + p2


differential.activity <- FindMarkers(
  object = mouse_brain,
  ident.1 = 'Pvalb',
  ident.2 = 'Sst',
  only.pos = TRUE,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

MotifPlot(
  object = mouse_brain,
  motifs = head(rownames(differential.activity)),
  assay = 'peaks'
)
```


# trajectory Monocle 3
```{r}
remotes::install_github('satijalab/seurat-wrappers')
BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
                       'limma', 'S4Vectors', 'SingleCellExperiment',
                       'SummarizedExperiment', 'batchelor', 'Matrix.utils'))
install.packages("devtools")
devtools::install_github('cole-trapnell-lab/leidenbase')
devtools::install_github('cole-trapnell-lab/monocle3')


library(SeuratWrappers)
library(monocle3)
library(Matrix)

combined.cds <- as.cell_data_set(combined)
combined.cds <- cluster_cells(cds = combined.cds, reduction_method = "umap2")
combined.cds <- learn_graph(combined.cds, use_partition = TRUE)

# order cells

#combined.cds <- order_cells(combined.cds, reduction_method = "UMAP", root_cells = hsc)
combined.cds <- order_cells(combined.cds, reduction_method = "UMAP")

# plot trajectories colored by pseudotime
plot_cells(
  cds = combined.cds,
  color_cells_by = "pseudotime",
  show_trajectory_graph = TRUE
)


combined <- AddMetaData(
  object = combined,
  metadata = combined.cds@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "all_pseudotime"
)

```


# footprinting
```{r}
library(motifmatchr)
library(JASPAR2020)
library(TFBSTools)
library(EnsDb.Mmusculus.v79)
library(BSgenome.Mmusculus.UCSC.mm10)

# extract position frequency matrices for the motifs
pwm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(species = 9606, all_versions = FALSE)
)

# add motif information
combined <- AddMotifs(combined, genome = BSgenome.Hsapiens.UCSC.hg38, pfm = pwm)

# gather the footprinting information for sets of motifs
combined <- Footprint(
  object = combined,
  motif.name = c("GATA2", "CEBPA", "EBF1"),
  genome = BSgenome.Mmusculus.UCSC.mm10
)

# plot the footprint data for each group of cells
p2 <- PlotFootprint(combined, features = c("GATA2", "CEBPA", "EBF1"))



```


```{r}
saveRDS(combined, "./BALF_combined.rds")
```

---
title: "MA10 downstream analysis"
author: "Jin Gyu Cheong"
date: "2022-09-15"
output: html_document
---


#Fig 1C,
```{r}
DimPlot(combined, reduction = "umap_ATAC", 
        group.by = "peaks_snn_res.0.2", label = T)
```




```{r}
DimPlot(combined, reduction = "umap_ATAC", 
        group.by = "celltype", label = T)
```


#Fig  1D,
```{r}
DimPlot(combined, reduction = "umap_ATAC", 
        group.by = "dataset", label = T)
```
#Fig  1F
subset macrophage
```{r}
macrophage <- subset(combined , peaks_snn_res.0.2 %in% c(0,1,4))


DefaultAssay(macrophage) <- "peaks"
macrophage <- RunTFIDF(macrophage)
macrophage <- FindTopFeatures(macrophage, min.cutoff = 20)
macrophage <- RunSVD(macrophage)
macrophage <- RunUMAP(macrophage, dims = 2:50, reduction = 'lsi', reduction.name = "ATAC_umap")

DimPlot(macrophage, reduction = "ATAC_umap") 
```

```{r}
DimPlot(macrophage, reduction = "ATAC_umap", group.by = "dataset") 
```

```{r}
FeaturePlot(macrophage, reduction = "ATAC_umap", cols = c("white","red"), split.by = "dataset", "MA1418.1", max.cutoff = 3, min.cutoff = -1)
FeaturePlot(macrophage, reduction = "ATAC_umap", cols = c("white","red"), split.by = "dataset", "MA0653.1", max.cutoff = 3, min.cutoff = -1)
FeaturePlot(macrophage, reduction = "ATAC_umap", cols = c("white","red"), split.by = "dataset", "MA0517.1", max.cutoff = 3, min.cutoff = -1)
```

#Fig  1G
```{r}
# differential chromvar activity anlaysis
diffTF <- FindMarkers(macrophage, assay = "chromvar", group.by = "dataset", ident.1 = "Recovered", ident.2 = "Naive")
diffTF$motif <- rownames(diffTF)
 names <- diffTF$motif %>% 
        TFBSTools::getMatrixByID(JASPAR2020, .) %>%
        lapply(., function (x) x@name)
diffTF$motifname <- names
```

volcano plot
```{r}
df <- diffTF
df$sig <- "non-significant"
df$sig[df$p_val < 0.05] <- "significant"

df$direction <- ""
df$direction[df$avg_log2FC > 0] <- "Recovered" 
df$direction[df$avg_log2FC < 0] <- "Naive" 

df$label <- ""
df$label[1:50] <- df$motifname[1:50]


ggplot(df, aes(x = avg_log2FC, y = -log(p_val), label = label))+
  geom_point()+
  geom_point(data = subset(df, sig =="significant"), mapping = aes(color = direction))+
  ylab("-log(pval)")+
  xlab("log2FC")+
  xlim(c(-max(df$avg_log2FC), max(df$avg_log2FC) + 1))+
  geom_text_repel(data = subset(df, df$direction== "Recovered") ,
                  size = 3,
                  max.overlaps = 100,
                  xlim = c(5,10),
                  hjust=0,
                  force= 0.5,
                  direction ="y",
                  segment.size = 0.2,
                  segment.color = "grey50")+
  geom_text_repel(data = subset(df, df$direction== "Naive") ,
                  size = 3,
                  max.overlaps = 1000,
                  xlim = c(-10,-5),
                  hjust=0,
                  force= 0.5,
                  direction ="y",
                  segment.size = 0.2,
                  segment.color = "grey50")+
  theme_classic()+
  scale_color_manual(values =
                       rev(wes_palette(n=2, name="Darjeeling1", type = "discrete")))+
  ggtitle("Naive <-> Recovered")

```

#Fig S1H,
```{r}
# Define the features (genes) you want to plot
features <- c("Pparg", "Mertk", "Itgax", "Apoe", "C1qa", "C1qb", 
              "Cd3g", "Cd8b1", "Themis", "Tcf7", "Il2ra", "Sell", 
              "Bcl2", "Itk", "Il7r", "Bach2", "Bank1", "Cd19", 
              "Cd83", "H2-M2", "Cd40", "Ccr7", "Epcam", "Cdh1", "Krt18")

# Generate the dot plot
dot_plot <- DotPlot(object = combined, features = features, group.by = "celltype", assay = "RNA") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  coord_flip()

dot_plot
```

#Fig S1 I-J
```{r}
# pefrom DEG analysis
deg <- FindMarkers(macrophage, assay = "RNA", group.by = "dataset", ident.1 = "Recovered", ident.2 = "Naive")
deg$gene_name <- rownames(deg)
```

script for running GSEA on DEG result
```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)


## RMA GSEA analysis

gseaobj <- function(df){
  tryCatch(
    {
      genelist <- df$avg_log2FC
      names(genelist) <- df$gene_name
      genelist <- na.omit(genelist)
      genelist <- sort(genelist, decreasing = T)
      
      gse <- gseGO(geneList=genelist, 
                   ont ="BP", 
                   keyType = "SYMBOL", 
                   minGSSize = 10, 
                   maxGSSize = 500,  # 500 genes seem too general
                   pvalueCutoff = 0.05, 
                   verbose = TRUE, 
                   OrgDb = "org.Hs.eg.db")
      gse
    }, error = function(e) {print("error")})
}

gseoplot <- function(x) {
  tryCatch(
    {
      dotplot(x, showCategory = 10, split = ".sign", font.size = 7) + facet_grid(.~.sign)
    }, error = function(e) {print("error")})
}


# perform GSEA on differential analysis  result
gsea <- gseaobj(df)
gsea_plt <- gseoplot(gsea)

gsea@result
```

volcano plots
```{r}
df <- deg
df$sig <- "non-significant"
df$sig[df$p_val < 0.05] <- "significant"

df$direction <- ""
df$direction[df$avg_log2FC > 0] <- "Recovered" 
df$direction[df$avg_log2FC < 0] <- "Naive" 

df$label <- ""
df$label[1:50] <- df$gene_name[1:50]


ggplot(df, aes(x = avg_log2FC, y = -log(p_val), label = label))+
  geom_point()+
  geom_point(data = subset(df, sig =="significant"), mapping = aes(color = direction))+
  ylab("-log(pval)")+
  xlab("log2FC")+
  xlim(c(-max(df$avg_log2FC), max(df$avg_log2FC) + 1))+
  geom_text_repel(data = subset(df, df$direction== "Recovered") ,
                  size = 3,
                  max.overlaps = 100,
                  xlim = c(5,10),
                  hjust=0,
                  force= 0.5,
                  direction ="y",
                  segment.size = 0.2,
                  segment.color = "grey50")+
  geom_text_repel(data = subset(df, df$direction== "Naive") ,
                  size = 3,
                  max.overlaps = 1000,
                  xlim = c(-10,-5),
                  hjust=0,
                  force= 0.5,
                  direction ="y",
                  segment.size = 0.2,
                  segment.color = "grey50")+
  theme_classic()+
  scale_color_manual(values =
                       rev(wes_palette(n=2, name="Darjeeling1", type = "discrete")))+
  ggtitle("Naive <-> Recovered")
```

#Fig S1L
```{r}

# seurat multiome dataset of our COVID study(10.1016/j.cell.2023.07.019) available in https://wcm.box.com/s/envl4ute8nfiaxq4vbw59vg6vz985gzn

covid <- readRDS("./covid_dataset.rds")
mild <- subset(covid, Status %in% c("Healthy", "Mild"))
mild <- subset(mild, MarkerAnnotations != "HSPC")

DefaultAssay(mild) <- "RNA"
mild<- NormalizeData(object = mild)
mild<- FindVariableFeatures(object = mild)
mild<- ScaleData(object = mild)
mild<- RunPCA(mild, verbose = FALSE)
mild <- RunHarmony(mild, "sample")
mild <- RunUMAP(mild, reduction = "harmony", dims = 1:20, reduction.name = "RNA_harmony")
mild <- FindNeighbors(mild, reduction = "harmony", dims = 1:20, graph.name = "RNA_harmony_snn")
mild <- FindClusters(mild, verbose = T, resolution = 0.8, graph.name = "RNA_harmony_snn")

DimPlot(mild, reduction = "RNA_harmony", group.by = "MarkerAnnotations", label = T , repel = T ) + NoLegend()
```

#Fig S1M-N
GO:0051607 : defesne response to virus
```{r}
# Subsetting the 'mild' object to include only specific marker annotations
mildmyeloid <- subset(mild, MarkerAnnotations %in% c("CD14 M", "CD16 M", "DC"))

# Define the Gene Ontology (GO) term of interest
id <- "GO:0051607"

# Load the biomaRt library for accessing Ensembl data
library(biomaRt)
ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")  # Set up the Ensembl dataset

# Retrieve gene data associated with the GO term from Ensembl
gene.data <- getBM(
  attributes = c("hgnc_symbol", "ensembl_transcript_id", "go_id"),  # Attributes to retrieve
  filters = "go",  # Filter by GO term
  values = id,  # GO term value
  mart = ensembl  # Mart object
)

# Filter the retrieved gene data to include only entries matching the GO term
gene.data_select <- gene.data %>% dplyr::filter(go_id == id)

# Extract unique human gene symbols associated with the GO term
gene.data_select_gene_human <- gene.data_select$hgnc_symbol %>% unique()
gene.data_select_gene_human  # Display the list of unique gene symbols

# Add a module score for the set of genes associated with the GO term
mildmyeloid <- AddModuleScore(
  object = mildmyeloid,
  features = list(gene.data_select_gene_human),  # List of gene sets to score
  assay = "RNA",  # Assay to use for the calculation
  name = "defense2virus",  # Prefix for the added score
  search = TRUE  # Search for the feature sets in the data
)

# Generate a violin plot for the module score, grouped by 'Status'
VlnPlot(mildmyeloid, "defense2virus1", group.by = "Status") +
  ggtitle("defense to virus in HSPC")

# Generate a violin plot for the module score, split by 'Status' and grouped by 'MarkerAnnotations'
VlnPlot(mildmyeloid, "defense2virus1", split.by = "Status", group.by = "MarkerAnnotations") +
  ggtitle("defense to virus in myeloid subset")

```

#Fig S1O
```{r}
# Find differentially expressed genes (DEGs) between "Recovered" and "Naive" in the "macrophage" dataset
deg <- FindMarkers(
  macrophage,  # Seurat object
  ident.1 = "Recovered",  # Group 1
  ident.2 = "Naive",  # Group 2
  group.by = "dataset",  # Grouping variable
  logfc.threshold = 0,  # Log fold change threshold for considering genes
  verbose = TRUE,  # Print progress
  only.pos = TRUE  # Only return positive markers
)

# Filter DEGs to include only those with avg_log2FC > 0.1 and adjusted p-value < 0.05
degname <- deg %>%
  dplyr::filter(avg_log2FC > 0.1) %>%
  dplyr::filter(p_val_adj < 0.05)
degname <- rownames(degname)  # Extract gene names

# Create a data frame with the filtered gene names
df <- data.frame(geneID = degname)

# Convert mouse gene symbols to human gene symbols using the mouse2human function
df$geneID <- lapply(df$geneID, function(x) {
  out <- mouse2human(x)  # Convert mouse gene symbol to human
  out$humanGene  # Extract human gene symbol
})

# Create a list of human gene symbols
gene_list <- df$geneID
names(gene_list) <- df$Description  # Assign descriptions as names (if applicable)

# Unlist the gene symbols
genes <- unlist(df$geneID)

# Add a module score for the set of DEGs in the "mildmyeloid" Seurat object
mildmyeloid <- AddModuleScore(
  object = mildmyeloid,
  features = list(genes),  # List of gene sets to score
  assay = "RNA",  # Assay to use for the calculation
  name = "MAC_deg",  # Prefix for the added score
  search = TRUE  # Search for the feature sets in the data
)

# Generate violin plots for the module score
VlnPlot(mildmyeloid, "MAC_deg1", group.by = "MarkerAnnotations", split.by = "Status", pt.size = 0) +
  ggtitle("MAC_deg1 in different MarkerAnnotations, split by Status")

VlnPlot(subset(mildmyeloid, MarkerAnnotations == "CD14 Monocytes"), "MAC_deg1", group.by = "Status", pt.size = 0, log = TRUE) +
  ggtitle("MAC_deg1 in CD14 Monocytes, grouped by Status")

RidgePlot(subset(mildmyeloid, MarkerAnnotations == "CD14 Monocytes"), "MAC_deg1", group.by = "Status") +
  ggtitle("Ridge plot of MAC_deg1 in CD14 Monocytes, grouped by Status")

# Additional violin plots for another gene set "macsdegsig_1"
VlnPlot(mildmyeloid, "macsdegsig_1", group.by = "MarkerAnnotations", split.by = "Status", pt.size = 0) +
  ggtitle("macsdegsig_1 in different MarkerAnnotations, split by Status")

VlnPlot(subset(mildmyeloid, MarkerAnnotations == "CD14 Monocytes"), "macsdegsig_1", group.by = "Status", pt.size = 0.1) +
  ggtitle("macsdegsig_1 in CD14 Monocytes, grouped by Status")

RidgePlot(subset(mildmyeloid, MarkerAnnotations == "CD14 Monocytes"), "macsdegsig_1", group.by = "Status") +
  ggtitle("Ridge plot of macsdegsig_1 in CD14 Monocytes, grouped by Status")

# Violin plot with additional annotations and statistical comparison
VlnPlot(subset(mildmyeloid, MarkerAnnotations == "CD14 Monocytes"), "macsdegsig_1", group.by = "Status", pt.size = 0.1) +
  ggtitle("DEG Rec vs Naive, \n in CD14 Monocytes") +
  stat_compare_means(ref.group = "Healthy", label = "p.format")  # Add statistical comparison
```

Fig S1P
```{r}
# Find differentially expressed transcription factors (TFs) between "Recovered" and "Naive" in CD14+ Monocytes
diffTF <- FindMarkers(
  subset(mildmyeloid, MarkerAnnotations == "CD14 M"),  # Subset to CD14+ Monocytes
  ident.1 = "Recovered",  # Group 1
  ident.2 = "Naive",  # Group 2
  group.by = "dataset",  # Grouping variable
  logfc.threshold = 0,  # Log fold change threshold for considering genes
  verbose = TRUE,  # Print progress
  only.pos = TRUE  # Only return positive markers
)

# Add motif names to the results
diffTF$motif <- rownames(diffTF)  # Assign row names to the 'motif' column

# Retrieve the motif names using JASPAR2020 database
names <- diffTF$motif %>%
  TFBSTools::getMatrixByID(JASPAR2020, .) %>%  # Get motif matrices by ID
  lapply(., function(x) x@name)  # Extract motif names

diffTF$motifname <- names  # Add the retrieved motif names to the 'motifname' column

# Create the bar plot for the differentially expressed transcription factors
ggplot(diffTF, aes(x = reorder(motifname, avg_log2FC), y = avg_log2FC, fill = log10_padj)) +
  geom_bar(stat = "identity") +  # Create a bar plot
  scale_fill_gradient(low = "blue", high = "green") +  # Set color gradient based on -log10(padj)
  coord_flip() +  # Flip coordinates for better readability
  theme_minimal() +  # Use a minimal theme for the plot
  labs(
    title = "CD14+ monocytes rec. vs. healthy",  # Add title to the plot
    x = "",  # X-axis label (left empty)
    y = "avg log2 fold change",  # Y-axis label
    fill = "-log10(padj)"  # Legend title for the fill color
  )
```


#Fig 2A
```{r}
# find marker genes of each cluster in the macrophage data
Idents(macrophage) <- "peaks_snn_res.0.2"
macrophag_mkr <- FindAllMarkers(macrophage, assay = "RNA")
macrophag_mkr
```

```{r}
# Define the features (genes) you want to plot
features <- c("Pparg", "Chil3", "Mertk", "Itgax", "Flt1", "Ear1", "Krt79",
              "Siglecf", "Car4", "Fabp4", "Marco", "Fabp5", "Apoe", 
              "C1qa", "C1qb", "C1qc", "Mafb", "Itgam", "S100a4", "Ccr2", 
              "Cd14", "Csf1r", "Fn1", "H2-Ab1", "S100a6")

# Generate the dot plot
dot_plot <- DotPlot(object = macrophage, features = features, group.by = "celltype", assay = "RNA") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_color_gradient(low = "lightpink", high = "red") +
  labs(title = "Macrophage signature genes", 
       x = "Cell Types", 
       y = "Genes")

dot_plot
```


#Fig S2A
```{r setup, include=FALSE}
features <- c("Apoe", "C1qc", "C1qb", "Psd3", "C1qa", "Spp1", "Dab2", "Tbc1d9",
              "Picalm", "Sash1", "Nav2", "Htr2c", "Cxcl2", "Adgrl3", "Kcnip4")

# Generate the dot plot
dot_plot <- DotPlot(object = macrophage, features = features, group.by = "celltype", assay = "RNA") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_color_gradient(low = "lightpink", high = "red") +
  labs(title = "Cluster marker genes", 
       x = "Cell Types", 
       y = "Genes")

dot_plot
```







